frontend/app/src/util/AppNavigation.ts

```diff
@@ -140,6 +152,131 @@ export class V1Strategy {
       ) ?? this.appPages[0]
     )
   }
+
+  clearPageElements(
+    _elements: AppRoot,
+    mainScriptHash: string,
+    sidebarElements: BlockNode | undefined
+  ): AppRoot {
+    return AppRoot.empty(mainScriptHash, false, sidebarElements)
+  }
+}
+
+export class StrategyV2 {
+  readonly appNav: AppNavigation
+
+  mainScriptHash: string | null
+
+  appPages: IAppPage[]
+
+  mainPage: IAppPage | null
+
+  hideSidebarNav: boolean | null
+
+  constructor(appNav: AppNavigation) {
+    this.appNav = appNav
+    this.mainScriptHash = null
+    this.appPages = []
+    this.mainPage = null
+    this.hideSidebarNav = null
+  }
+
+  handleNewSession(newSession: NewSession): MaybeStateUpdate {
+    this.mainScriptHash = newSession.mainScriptHash
+    // Initialize to the config value if provided
+    if (this.hideSidebarNav === null) {
+      this.hideSidebarNav = newSession.config?.hideSidebarNav ?? null
+    }
+
+    // We do not know the page name, so use an empty string version
+    document.title = getTitle("")
+
+    return [{ hideSidebarNav: this.hideSidebarNav ?? false }, () => {}]
+  }
+
+  handlePagesChanged(_pagesChangedMsg: PagesChanged): MaybeStateUpdate {
+    // This message does not apply to V2
+    return undefined
+  }
+
+  handlePageNotFound(pageNotFound: PageNotFound): MaybeStateUpdate {
+    const { pageName } = pageNotFound
+    this.appNav.onPageNotFound(pageName)
+
+    return [
+      { currentPageScriptHash: this.mainScriptHash ?? "" },
+      () => {
+        this.appNav.hostCommunicationMgr.sendMessageToHost({
+          type: "SET_CURRENT_PAGE_NAME",
+          currentPageName: "",
+          currentPageScriptHash: this.mainScriptHash ?? "",
+        })
+      },
+    ]
+  }
+
+  handleNavigation(navigationMsg: Navigation): MaybeStateUpdate {
+    const { sections, position, appPages } = navigationMsg
+
+    this.appPages = appPages
+    this.hideSidebarNav = position === "hidden"
+
+    const currentPage = appPages.find(
+      p => p.pageScriptHash === navigationMsg.pageScriptHash
```

Comment:
nit: should we extract `pageScriptHash` from `navigationMsg` in the destructor already used above to be used here instead of accessing it from the object?